orc8r/cloud/go/obsidian/access/rest_middleware.go

```diff
@@ -100,3 +106,82 @@ func makeErr(decorate logDecorator, status int, errFmt string, errArgs ...interf
 	glog.V(1).Infof("REST middleware (obsidian) rejected request: %s", decorate(errFmt, errArgs...))
 	return echo.NewHTTPError(status, fmt.Sprintf(errFmt, errArgs...))
 }
+
+// TokenMiddleware parses the <username>:<token> from the header, validates the token,
+// then checks if the request is within the specified permissions granted to the user.
+func TokenMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		req := c.Request()
+
+		// Skip middleware if request when there is no security requirement
+		// for an endpoint
+		noBasicAuthEndpoints := []string{"/magma/v1/user/login"}
+		for _, endpoint := range noBasicAuthEndpoints {
+			if req.RequestURI == endpoint {
+				return next(c)
+			}
+		}
+
+		username, token, ok := req.BasicAuth()
+		if !ok {
+			return echo.NewHTTPError(http.StatusBadRequest, "failed to parse basic auth header")
+		}
+
+		// Return early without going to the certifier service if token is invalid
+		if err := certifier.ValidateToken(token); err != nil {
+			return echo.NewHTTPError(http.StatusBadRequest, err)
+		}
+		resourceType, resourceVal := getResource(c)
+		resource := &certprotos.Resource{
+			Action:       getRequestAction(req, nil),
+			ResourceType: resourceType,
+			Resource:     resourceVal,
+		}
+		getPDReq := &certprotos.GetPolicyDecisionRequest{
+			Username: username,
+			Token:    token,
+			Resource: resource,
+		}
+		pd, err := certifier.GetPolicyDecision(req.Context(), getPDReq)
+		if err != nil {
+			return obsidian.MakeHTTPError(err, http.StatusInternalServerError)
+		}
+		if pd.Effect == certprotos.Effect_DENY {
+			return echo.NewHTTPError(http.StatusForbidden, "not authorized to view resource")
+		}
+		if next != nil {
+			glog.V(4).Info("Token middleware successfully verified permissions. Sending request to the next middleware.")
+			return next(c)
+		}
+
+		return nil
+	}
+}
+
+// getRequestType returns the required request permission (READ, WRITE
+// or READ+WRITE) corresponding to the request method.
+func getRequestAction(req *http.Request, decorate logDecorator) certprotos.Action {
+	switch req.Method {
+	case "GET", "HEAD":
+		return certprotos.Action_READ
+	case "PUT", "POST", "DELETE":
+		return certprotos.Action_WRITE
+	default:
+		glog.Info(decorate("Unclassified HTTP method '%s', defaulting to read+write requested permissions", req.Method))
+		return certprotos.Action_READ | certprotos.Action_WRITE
+	}
+}
+
+func getResource(c echo.Context) (certprotos.ResourceType, string) {
+	networkIDStr := strings.ToLower(certprotos.ResourceType_NETWORK_ID.String())
+	tenantIDStr := strings.ToLower(certprotos.ResourceType_TENANT_ID.String())
+	for _, p := range c.ParamNames() {
+		switch p {
+		case networkIDStr:
+			return certprotos.ResourceType_NETWORK_ID, c.Param(networkIDStr)
+		case tenantIDStr:
+			return certprotos.ResourceType_TENANT_ID, c.Param(tenantIDStr)
+		}
+	}
+	return certprotos.ResourceType_URI, c.Request().RequestURI
+}
```

Comment:
Sooo to handle scoping by network/tenant ID, I think this is probably the best way to handle those endpoints? But this also means that when creating a policy, the operator needs to specify separate policies for handling the endpoints with network/tenant ID.

Reply:
Ig I am not sure how the certifier can determine that something in the path is a resource rather than just a normal path component. Would we not have to go through some registry of all endpoints to first check if an endpoint takes in network/tenant ID in its path param then do a bunch of parsing to actually get the id? Currently I'm getting this info from echo.Context, but gRPC it can only get context.Context, which I don't think retains the param info, but maybe I'm wrong, but yeah we can talk more tmrw.